bench_test.go

```diff
@@ -0,0 +1,179 @@
+package dhstore_test
+
+import (
+	"bytes"
+	"context"
+	crand "crypto/rand"
+	"encoding/json"
+	"io"
+	"log"
+	"math"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/ipni/dhstore"
+	"github.com/ipni/dhstore/metrics"
+	"github.com/multiformats/go-multihash"
+	"github.com/stretchr/testify/require"
+)
+
+func BenchmarkDHStore_PutMultihashes_0(b *testing.B) {
+	benchmarkPut(b, 300, 1000, 200)
+}
+
+func BenchmarkDHStore_GetMultihashes_0(b *testing.B) {
+	benchmarkGet(b, 500, 1000, 200)
+}
+
+func benchmarkPut(b *testing.B, chunks, chunkSize, vkLen int) {
+	b.SetBytes(int64(34 + vkLen))
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	dir := b.TempDir()
+
+	store, metr, server := createServer(b, dir)
+	defer store.Close()
+	defer metr.Shutdown(context.Background())
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mhs := randomMultihashes(b, chunks*chunkSize)
+			putMultihashes(b, mhs, chunks, vkLen, server, true)
+		}
+	})
+	b.StopTimer()
+}
+
+func benchmarkGet(b *testing.B, chunks, chunkSize, vkLen int) {
+	b.ResetTimer()
+
+	dir := b.TempDir()
+
+	store, metr, server := createServer(b, dir)
+	defer store.Close()
+	defer metr.Shutdown(context.Background())
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mhs := randomMultihashes(b, chunks*chunkSize)
+			putMultihashes(b, mhs, chunks, vkLen, server, false)
+			rand.Shuffle(len(mhs), func(i, j int) {
+				mhs[i], mhs[j] = mhs[j], mhs[i]
+			})
```

Comment:
* Ditto re moving out setup. Bencharking get should only benchmark get.
* Make sure random shuffling is reproducible and deterministic given a seed.

Reply:
Move shuffling out of the benchmark?